### DISCUSSION TITLE
Bulk Invert images, Remove background, Add Border, Resize, Glow effect and renaming in Windows

### DISCUSSION THREAD

**Discussion by Archive-d** [ImageMagick/ImageMagick#5162] [Help]
OS: Windows 10 IM 7.1 Q16

Hello everyone,
I am kind of new to this and also coding in general so please bear with me.

I am Looking for a way to edit multiple images while keeping the names intact.

Example
**Invert**

![AirBnb.jpg](https://user-images.githubusercontent.com/76863644/169281269-47471467-dbc8-4b89-9a6f-d13eef6ac569.png)
![AirBnb.jpg](https://user-images.githubusercontent.com/76863644/169281302-031cd214-8972-4dcc-a01b-9b2f7f8ecd04.png)

**Remove Background with both white or black as BG**

![AirBnb](https://user-images.githubusercontent.com/76863644/169281568-57add073-294e-456a-a85c-0fe8cda27b8d.png)
![AirBnb](https://user-images.githubusercontent.com/76863644/169281648-a6c636fd-281a-45f6-823e-5dcdb1cedc50.png)

**Add Border in Png images**

![AirBnb](https://user-images.githubusercontent.com/76863644/169281887-c9f24822-d375-4f5a-9d81-90c6f660e785.png)
![AirBnbx](https://user-images.githubusercontent.com/76863644/169281896-2d77c4a9-6886-4632-abfa-d04146fad56d.png)

**If Possible add a glow effect** 

![AirBnb-BLUE](https://user-images.githubusercontent.com/76863644/169282167-62090dbb-6bdc-4230-a83d-27f0d652b2c1.png)
![AirBnbx-BLUE](https://user-images.githubusercontent.com/76863644/169282186-705e655f-ae7a-489f-95f0-1e59f59e1006.png)
![AirBnb-Lime](https://user-images.githubusercontent.com/76863644/169282291-91d946bf-5650-45d2-a29e-9eeed8066636.png)

**Resize images: Resize images to a specific size, Ex: 958x958 to 1080x1080.**

P.s I have come across scripts that do pretty much everything on here, but I do not know how to perform on bulk. Also keep the names, as in, Airbnb.png to be saved as Airbnb.png or Airbnb_glow.png
**All requests are individual, I do not want one single process for all combined**

Would be amazing if the images get saved to a separate folder but I am fine with overwriting them if required.

Anything on this would be really helpful. _/\__/\_
Thank you in advance.

**Comment by snibgo** [comment]
> I am Looking for a way to edit multiple images while keeping the names intact.

I suggest a "for" loop, like this (Windows BAT syntax):
```
for %%A in (*.png) do magick %%A {some_operation} outdir\%%~nA.png
```
... where "{some_operation}" is any IM operation(s) such as "-negate".

For help on "for", type "help for" at the command line. See many of my answers on this forum.

For help on IM operations, see https://imagemagick.org/script/command-line-options.php

See also http://imagemagick.org/Usage/ and http://imagemagick.org/Usage/basics/

**Reply by Archive-d** [reply]
Hey, Thank you for the quick reply.
could you help me with commands for Removing black/White Backgrounds? and for the Glow Effect.

Thank you again _/\_

**Reply by Archive-d** [reply]
Hey,
It worked really well.
I am having a small issue, If the script comes across any images with " " i.e space in their names I get the error

`No such file or directory @ error/blob.c/OpenBlob/3533.
magick: no decode delegate for this image format `' @ error/constitute.c/ReadImage/738.`

Is there a way to make this work or should I rename all my files and remove any spaces?

**Reply by snibgo** [reply]
Put the filenames in quotes:
```
for %%A in (*.png) do magick "%%A" {some_operation} "outdir\%%~nA.png"
```

**Reply by Archive-d** [reply]
Sorry to bother again,

When trying to use the resize command

`for %%A in (D:\Icons\RainbowWedoAes\Black\ICONSBlack\*) do magick convert "%%A" -resize 90% -channel RGB -negate "D:\Icons\MyIcons\%%~nA.png"`

I get this error

`convert: invalid argument for option '-resize': 90\Icons\MyIcons\Amazon @ error/convert.c/ConvertImageCommand/2655.`

I tried using `-resize 90% -quality 100%` and it worked but the resulting image is only of 90x90 and not 90% reduction of the real image or 90% of the real image.

**Comment by snibgo** [comment]
> ... commands for Removing black/White Backgrounds?

What do you mean by "removing" and "background?

You can make white pixels transparent like this:
```
magick in.ext -transparent White out.ext
```

>... and for the Glow Effect.

I wrote a page on [Soft focus: spreading the light](http://im.snibgo.com/softfoc.htm).

**Reply by Archive-d** [reply]
Hey,
I tried the soft focus script and some other on the net. They kind of blur and spread the image, doesn't look much like a glow.

Here's how I do it in photoshop, which I think could work with
**1. I have a black png image,**
![Airbnb](https://user-images.githubusercontent.com/76863644/169784150-a958adb4-686c-49c3-bfd7-64ba6ee890c3.png)

**2. Fill/ overlay with desired color, Lets say pink (Hash value #ff00ff) (RGB Value: 255, 0, 255). Is there a fill command on IM?**
![airbnb-pink](https://user-images.githubusercontent.com/76863644/169784312-1f54c577-b75d-4aa7-86e7-ad2044654905.gif)

**3. Clone the same image and apply a blur/ gaussian blur**
which I figured can be achieved though
```magick convert image.png -filter Gaussian -resize 50%  -define filter:sigma=2.5 -resize 200%  out.png```
or
`magick convert image.png -resize 25% -blur 0x8 -resize 400% out.png`

_The resize is done to speed up the process._ 

![out](https://user-images.githubusercontent.com/76863644/169791165-5099113a-048e-497d-bf87-0a2011debb82.png)

**4. Merge Images from _step2_ and _step3_ onto a plain black image**

![glow](https://user-images.githubusercontent.com/76863644/169791754-e445632d-af04-44fa-af3b-955b31534498.png)

**I need your help in simplifying this process. and achieving it in a single command** or 
_maybe a single command till step 3 and merging the images in one,_ 
![blur](https://user-images.githubusercontent.com/76863644/169802793-29143440-7789-45cf-a8a5-8356897bc470.png)

Then a separate command to Overlay that on a black background; which I can now achieve using "for %%A" command through your help on my previous question.

Thank you _/ \_

**Reply by fmw42** [reply]
You can make a glow effect using -shadow with zero X and Y components. Note that your image has all the information in the alpha channel and a black RGB component.

IM 7 Unix syntax:

```
magick paperclip.png \
\( +clone -background "#ff00ff" -shadow 100x30+0+0 \) +swap \
-background none   -layers merge  +repage  \
-fuzz 99% -fill "#ff00ff" -opaque black \
-background black -flatten \
paperclip_glow.png

```

![paperclip_glow](https://user-images.githubusercontent.com/13458218/169875588-eeef8663-243e-4dc5-b9e5-faeb856e6ffa.png)

**Reply by fmw42** [reply]
Here is a better version

```
magick paperclip.png -set option:dims "%wx%h" +write mpr:img \
\( +clone -background "#ff00ff" -shadow 100x30+0+0 \) +swap \
-background none -layers merge +repage \
-gravity center -crop "%[dims]+0+0" +repage \
\( mpr:img -channel rgb -fill "#ff00ff" -colorize 100% +channel \) \
-compose over -composite \
-background black -flatten \
paperclip_glow.png

```

![paperclip_glow](https://user-images.githubusercontent.com/13458218/169924106-73bd2b03-d489-4f44-a49f-5ef18cbc6083.png)

**Reply by Archive-d** [reply]
Hello,
Thank you for your response
I tried using both the scripts you provided and I got the same result i.e,

![paperclip_glow](https://user-images.githubusercontent.com/76863644/169974718-b84398d6-7188-4047-afaf-792576fbf15e.png)

I tried doing something myself 

`for %%A in (D:\Icons\Trial\FG\*) do magick "%%A" -fuzz 99%% -fill "#ff00ff" -colorize 100 ( +clone -resize 25%% -blur 15x15 -resize 400%% +repage ) +swap ( +clone -background black ) +swap -layers merge "D:\Icons\Trial\%%~nA.png"
`

While this works and is giving me the desired result, the edges are jagged

![paperclip](https://user-images.githubusercontent.com/76863644/169975229-f523a721-eb1f-46ab-9d5a-096666361348.png)

Anyway I can clear the edges? 
Also, I am having to use Double %% for -resize and -fuzz values. Is this right or could there be an error?

OS. IM 7 on windows 10

Thank you again!

**Reply by fmw42** [reply]
Yes, in Windows .bat files % have to be doubled. In Windows, parenthesis do not need escaping. Also end of lines are specified by ^ rather than \.

**Reply by fmw42** [reply]
In Windows DOS, this version should have better edges; otherwise, one would have to do some anti-aliasing via -blur etc.

```
magick paperclip.png -set option:dims "%wx%h" +write mpr:img ^
( +clone -background "#ff00ff" -shadow 100x30+0+0 ) +swap ^
-background none -layers merge +repage ^
-gravity center -crop "%[dims]+0+0" +repage ^
( mpr:img -channel rgb -fill "#ff00ff" -colorize 100% +channel ) ^
-compose over -composite ^
-background black -flatten ^
paperclip_glow.png

```

In a .bat script, double the %

**Reply by Archive-d** [reply]
Hi,
Sorry to bother again,  
How to I run it in windows DOS? I tried searching and it is an OS?

Is there a way I can use your version of the script? The edges are really bothering, I have tried all possible methods, including -unsharp -alpha extract etc etc.

Even when removing background

![out](https://user-images.githubusercontent.com/76863644/170273364-56a3e252-0604-4d32-91e6-319d6d18b665.png)

If I increase the fuzz values ~90%  the white edges vanish but the edges are jagged extremely

![out2](https://user-images.githubusercontent.com/76863644/170273628-5ef06cbd-6bce-4861-bcf7-4f9d8830d68f.png)
 
Any other way I could smooth the edges? Or if you can guide me towards any other program where I can do this for bulk?

Any help will be extremely appreciated.

Thank you.

**Reply by fmw42** [reply]
Try the version above that I said was a better method that does not involve -fuzz. I wrote the DOS version for you there. Did you not see that answer?

**Reply by Archive-d** [reply]
Hello, 
I tried using the above version with ^ but the edges were still jagged. 

And if possible, I need your help in removing the background. 
I tried `-transparent white` but it leaves white edges around the subject, which I tried clearing through -fuzz which ruins the edges.

**Reply by fmw42** [reply]
Try this

```
magick 170273628-5ef06cbd-6bce-4861-bcf7-4f9d8830d68f.png ^
-set option:dims "%wx%h" ^
-alpha extract -alpha copy +write mpr:img ^
( +clone -background "#ff00ff" -shadow 100x30+0+0 ) +swap ^
-background none -layers merge +repage ^
-gravity center -crop "%[dims]+0+0" +repage ^
( mpr:img -channel rgb -fill "#ff00ff" -colorize 100% +channel ) ^
-compose over -composite ^
-background black -flatten ^
170273628-5ef06cbd-6bce-4861-bcf7-4f9d8830d68f_glow.png

```
![170273628-5ef06cbd-6bce-4861-bcf7-4f9d8830d68f_glow](https://user-images.githubusercontent.com/13458218/170303854-c7d53eeb-46d0-4f55-bece-e3fcf9ffb08a.png)

**Reply by fmw42** [reply]
Or if you want to smooth (anti-alias) the edges, add some blur and leveling.

```
magick 170273628-5ef06cbd-6bce-4861-bcf7-4f9d8830d68f.png ^
-set option:dims "%wx%h" ^
-alpha extract -alpha copy ^
-blur 0x2 -level 50x100% +write mpr:img \
( +clone -background "#ff00ff" -shadow 100x30+0+0 ) +swap ^
-background none -layers merge +repage ^
-gravity center -crop "%[dims]+0+0" +repage ^
( mpr:img -channel rgb -fill "#ff00ff" -colorize 100% +channel ) ^
-compose over -composite ^
-background black -flatten ^
170273628-5ef06cbd-6bce-4861-bcf7-4f9d8830d68f_glow.png

```
![170273628-5ef06cbd-6bce-4861-bcf7-4f9d8830d68f_glow2](https://user-images.githubusercontent.com/13458218/170304518-5e5506d7-483d-41bd-afcf-28feee1530db.png)

**Reply by fmw42** [reply]
Here is another version that may be a bit more efficient.

```
magick 170273628-5ef06cbd-6bce-4861-bcf7-4f9d8830d68f.png ^
-set option:dims "%wx%h" ^
-alpha extract ^
( +clone -fill "#ff00ff" -colorize 100 ) ^
+swap -compose copy_opacity -composite -compose over +write mpr:img ^
( +clone -background "#ff00ff" -shadow 100x30+0+0 ) +swap ^
-background black -layers merge +repage ^
-gravity center -crop "%[dims]+0+0" +repage ^
170273628-5ef06cbd-6bce-4861-bcf7-4f9d8830d68f_glow4.png
```

![170273628-5ef06cbd-6bce-4861-bcf7-4f9d8830d68f_glow4](https://user-images.githubusercontent.com/13458218/170342229-b023109a-0aa6-48de-8318-81b33389de30.png)

**Comment by snibgo** [comment]
> How to I run it in windows DOS? I tried searching and it is an OS?

"Windows DOS" is an out-of-date nickname for "Windows Command Console", aka "Windows CMD", aka "Command Window".
